File: java/src/main/java/org/sparky/model/builders/VariableBuilder.java
package org.sparky.model.builders;

import org.sparky.model.Bit;
import org.sparky.model.KeyRule;
import org.sparky.model.Rule;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by lloyd on 17/10/2015.
 */
public class VariableBuilder implements EvaluatableBuilder {

    private final List<Bit> bits = new ArrayList<>();

    @Override
    public VariableBuilder withBit(Bit bit) {
        bits.add(bit);
        return this;
    }

    @Override
    public Rule build() {
        return new KeyRule(bits);
    }

    public String buildAsColumnName(){
        if(bits.size() != 1){
            throw new IllegalStateException("A TableColumn Variable should only have a single bit");
        }
        if(!(bits.get(0) instanceof Rule.KeyBit)){
            throw new IllegalStateException("A TableColumn Variable should only have a single KeyBit");
        }
        return ((Rule.KeyBit)bits.get(0)).getKey();
    }
}

File: net/Sparky/Model/Builder/VariableBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sparky.Model.Builder
{
    public class VariableBuilder : IEvaluatableBuilder
    {
        private readonly List<IPartial> _bits = new List<IPartial>();

        public string BuildAsColumnName()
        {
            if(_bits.Count != 1)
            {
                throw new InvalidOperationException("This Variable cannot be a Column Name because it is made from multiple Partials");
            }
            if(!(_bits[0] is SimpleKeyPartial))
            {
                throw new InvalidOperationException("This Variable cannot be a Column Name because its only Partial is not a SimpleKeyPartial");
            }
            return (_bits[0] as SimpleKeyPartial).GetKey();
        }

        public IEvaluatableBuilder WithPartial(IPartial partial)
        {
            _bits.Add(partial);
            return this;
        }

        public Rule Build()
        {
            return new KeyRule(_bits);
        }
    }
}
